### PR TITLE
Add "arguments" deserialization tests for "script.callFunction"

### PR DESCRIPTION
**Pull request by sadym-chromium**
* Added tests for arguments deserialization in `script.callFunction`

### CODE REVIEW

**Comment by lutien** [comment]
Looks perfect to me!
Tagging @whimboo and @juliandescottes, if you also want to have a look

File: webdriver/tests/bidi/script/call_function/arguments.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        {"type": "number", "value": 3},
+        {"type": "number", "value": 1.4},
+        {"type": "boolean", "value": True},
+        {"type": "boolean", "value": False},
+        {"type": "bigint", "value": "42"},
+    ],
+)
+async def test_primitive_values(bidi_session, top_context, argument):
+    result = await bidi_session.script.call_function(
+        function_declaration=f"(arg) => arg",
```
Comment (whimboo) [inline webdriver/tests/bidi/script/call_function/arguments.py:27]: So the problem here is that we actually do not compare the result of the deserialization but a complete round trip including deserialization and serialization. In case of the serialization also be broken we wouldn't detect a failure for the deserialization.

As such we would need an assertion for a type check as part of the function declaration. If that passes the value also needs to be checked. Then the test will pass if no `exceptionDetails` are part of the return value for the `script.callFunction` call.

The same actually also applies to the other test.

**Review by whimboo — DISMISSED** [review]

**Comment by sadym-chromium** [comment]
@whimboo, @juliandescottes?

File: webdriver/tests/bidi/script/call_function/arguments.py (hunk 1 above)
Comment (juliandescottes) [reply webdriver/tests/bidi/script/call_function/arguments.py:27]: @sadym-chromium fyi whimboo is off for a few weeks, but I agree with his feedback here. Ideally this should only assert deserialization. I received a ping for this review today but I didn't see a new patch. I'm assuming the ping was sent before seeing the comment above.

Comment (sadym-chromium) [reply webdriver/tests/bidi/script/call_function/arguments.py]: Yeah, I pinged before I saw the comment.

Ideally, we want to test only deserialization, but comparing sets, arrays, maps and objects on the backend is quite messy. So having we tested serialization, I guess it's acceptable trade-off just to check the types of non-pimitives on the server side, and check the "deserialized-serialized" values on the client side.

Comment (sadym-chromium) [reply webdriver/tests/bidi/script/call_function/arguments.py]: @juliandescottes

File: webdriver/tests/bidi/script/call_function/arguments.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        f"        throw Error(\"Argument type should be {expected_type}, but was \"+"
+        f"            Object.prototype.toString.call(arg));"
+        f"    return arg;"
+        f"}}",
+        arguments=[argument],
+        await_promise=False,
+        target=ContextTarget(top_context["context"]),
+    )
+
+    assert result == argument
```
Comment (juliandescottes) [inline webdriver/tests/bidi/script/call_function/arguments.py:116]: I think we should rather check type & value here. The spec does not guarantee that a LocalValue and a RemoteValue can be compared safely.

File: webdriver/tests/bidi/script/call_function/arguments.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        ({"type": "bigint", "value": "42"}, "42n"),
+    ],
+)
+async def test_primitive_values(bidi_session, top_context, argument, expected):
+    result = await bidi_session.script.call_function(
+        function_declaration=
+        f"(arg) => {{ "
+        f"    if(arg!=={expected}) "
+        f"        throw Error(\"Argument type should be {expected}, but was \"+arg);"
+        f"    return arg;"
```
Comment (juliandescottes) [inline webdriver/tests/bidi/script/call_function/arguments.py:30]: Let's just return `arg===expected` and check that `result` is `True`?

File: webdriver/tests/bidi/script/call_function/arguments.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+@pytest.mark.asyncio
+async def test_nan(bidi_session, top_context):
+    nan_remote_value = {"type": "number", "value": "NaN"}
+    result = await bidi_session.script.call_function(
+        function_declaration=
+        f"(arg) => {{ "
+        f"    if(!isNaN(arg)) "
+        f"        throw Error(\"Argument type should be 'NaN', but was \"+arg);"
+        f"    return arg;"
```
Comment (juliandescottes) [inline webdriver/tests/bidi/script/call_function/arguments.py:48]: Similar comment, return `isNaN(arg)` and check if `result` is `True`.

**Review by juliandescottes — CHANGES_REQUESTED** [review]
Thanks for the update Maksim! The compromise sounds fine, but asserting arrays, maps, sets (...) in JS would not make the test that hard to follow IMO. Back to you for a few comments, I'll take one last look with those addressed.

File: webdriver/tests/bidi/script/call_function/arguments.py (hunk 3 above)
Comment (sadym-chromium) [reply webdriver/tests/bidi/script/call_function/arguments.py:30]: Having an exception with a meaningful message thrown makes debugging way easier, as the error message contains the meaningful message, rather than "expected `true`, but was `false`".

File: webdriver/tests/bidi/script/call_function/arguments.py (hunk 4 above)
Comment (sadym-chromium) [reply webdriver/tests/bidi/script/call_function/arguments.py:48]: Answered here: https://github.com/web-platform-tests/wpt/pull/34830#discussion_r927493703

File: webdriver/tests/bidi/script/call_function/arguments.py (hunk 2 above)
Comment (sadym-chromium) [reply webdriver/tests/bidi/script/call_function/arguments.py]: Switched to `recursive_compare`.

File: webdriver/tests/bidi/script/call_function/arguments.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    ],
+)
+async def test_primitive_values(bidi_session, top_context, argument, expected):
+    result = await bidi_session.script.call_function(
+        function_declaration=
+        f"(arg) => {{ "
+        f"    if(arg!=={expected}) "
+        f"        throw Error(\"Argument type should be {expected}, but was \"+arg);"
+        f"    return arg;"
+        f"}}",
```
[the remark this answers is not in the document]
Comment (sadym-chromium) [reply webdriver/tests/bidi/script/call_function/arguments.py]: done

File: webdriver/tests/bidi/script/call_function/arguments.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+            if(!isNaN(arg))
+                throw Error("Argument should be 'NaN', but was "+arg);
+            return arg;
+        }}""",
+        arguments=[nan_remote_value],
+        await_promise=False,
+        target=ContextTarget(top_context["context"]),
+    )
+
+    assert result == nan_remote_value
```
Comment (juliandescottes) [inline webdriver/tests/bidi/script/call_function/arguments.py:56]: Can we use recursive compare here as well?

**Review by juliandescottes — APPROVED** [review]
Thanks!

File: webdriver/tests/bidi/script/call_function/arguments.py (hunk 6 above)
Comment (sadym-chromium) [reply webdriver/tests/bidi/script/call_function/arguments.py:56]: makes sense. Done
